prometheus-client/benchmarks/Main.hs

```diff
@@ -1,44 +1,147 @@
+{-# language FlexibleContexts #-}
+
+module Main (main) where
+
 import Prometheus
 
 import Control.Monad
+import Criterion
 import Criterion.Main
+import Data.Foldable (for_)
 import System.Random
 
+withMetric m =
+  envWithCleanup
+    (registerIO m)
+    (const unregisterAll)
+
+withCounter =
+  withMetric (counter (Info "a" "b"))
+
+withGauge =
+  withMetric (gauge (Info "a" "b"))
+
+withSummary quantiles =
+  withMetric (summary (Info "a" "b") quantiles)
+
+withHistogram buckets =
+  withMetric (histogram (Info "a" "b") buckets)
 
 main :: IO ()
-main = defaultMain [
-        bgroup "incCounter" $ expandBenches incCounterThenCollect
-    ,   bgroup "withLabelIncCounter" $ expandBenches withLabelIncCounter
-    ,   bgroup "addGauge" $ expandBenches $ withGaugeThenCollect (addGauge 47.0)
-    ,   bgroup "subGauge" $ expandBenches $ withGaugeThenCollect (subGauge 47.0)
-    ,   bgroup "setGauge" $ expandBenches $ withGaugeThenCollect (setGauge 47.0)
-    ,   bgroup "observe" $ expandBenches observeThenCollect
+main =
+  defaultMain
+    [ benchCounter
+    , benchGauge
+    , benchSummary
+    , benchHistogram
+    , benchExport
+    ]
+
+
+-- Counter benchmarks
+
+
+benchCounter =
+  withCounter $ \counter ->
+    bgroup "Counter"
+      [ benchIncCounter counter
+      , benchAddCounter counter
+      , benchAddDurationToCounter counter
+      ]
+
+benchIncCounter testCounter =
+  bench "incCounter" $ whnfIO (incCounter testCounter)
+
+benchAddCounter testCounter =
+  bench "addCounter" $ whnfIO (addCounter 50 testCounter)
+
+benchAddDurationToCounter testCounter =
+  bench "addDurationToCounter" $ whnfIO (addDurationToCounter (return ()) testCounter)
+
+
+
+-- Gauge benchmarks
+
+
+benchGauge =
+  withGauge $ \gauge ->
+    bgroup "Gauge"
+      [ benchIncGauge gauge
+      , benchAddGauge gauge
+      , benchSubGauge gauge
+      , benchSetGaugeToDuration gauge
+      ]
+
```

Comment:
Is this benchmarking a single increment? Is it possible for criterion to get a meaningful measurement from such a short operation?

I feel like I remember needing to repeat operations many times in order to get a duration large enough that I could see the needle move when optimizing code.